### PR TITLE
24 allow for dynamic plugins

### PR DESCRIPTION
**Pull request by cfirth-nasa**
Add support for external plugins to be specified by name and path in config INI's and imported as Plugins in data_driven_learning.py

### CODE REVIEW

File: onair/config/default_config.ini [hunk 1]
```diff
@@ -6,6 +6,7 @@ TelemetryFiles = ['700_crash_to_earth_1.csv']
 ParserFileName = csv_parser
 ParserName = CSV
 SimName = CSV
+PluginList = {'generic_plugin':'plugins/generic/generic_plugin.py'} # format: {plugin_name : plugin_path}
```
Comment (Evana13G) [inline onair/config/default_config.ini:9]: Do we have an area where we can put an assertion statement?

File: onair/src/data_driven_components/data_driven_learning.py [hunk 2]
```diff
@@ -10,17 +10,22 @@
...
 class DataDrivenLearning:
-    def __init__(self, headers, _ai_plugins:list=[]):
+    def __init__(self, headers, _ai_plugins={}):
         assert(len(headers)>0)
         self.headers = headers
-        self.ai_constructs = [
-            importlib.import_module('onair.src.data_driven_components.' + plugin_name + '.' + f'{plugin_name}_plugin').Plugin(plugin_name, headers) for plugin_name in _ai_plugins
-        ]
+        self.ai_constructs = []
+
```
Comment (Evana13G) [inline onair/src/data_driven_components/data_driven_learning.py:22]: assertion here, or earlier. Also, I'd remove this empty space.

File: onair/src/reasoning/agent.py [hunk 3]
```diff
@@ -15,9 +15,9 @@
 from ..reasoning.diagnosis import Diagnosis
 
 class Agent:
-    def __init__(self, vehicle):
+    def __init__(self, plugin_list, vehicle):
```
Comment (Evana13G) [inline onair/src/reasoning/agent.py:18]: I'd swap the order (most significant to least significant)

Comment (Evana13G) [reply onair/src/reasoning/agent.py:18]: Also, follows the instantiation order

File: onair/src/run_scripts/execution_engine.py [hunk 4]
```diff
@@ -94,6 +98,8 @@ def parse_configs(self, config_filepath):
             self.benchmarkIndices = config['DEFAULT']['BenchmarkIndices']
         except:
             pass
+        ## Sort Data: Plugins
```
Comment (Evana13G) [inline onair/src/run_scripts/execution_engine.py:101]: is sorting happening? I think this function turns a string into list (with extra processing of tokens , beyond pythons built in "list()").

File: onair/src/run_scripts/sim.py [hunk 5]
```diff
@@ -39,7 +39,7 @@ def __init__(self, simType, parsedData, SBN_Flag):
             
         else:
             self.simData = DataSource(parsedData.get_sim_data())
-        self.agent = Agent(vehicle)
+        self.agent = Agent(plugin_list,vehicle)
```
Comment (Evana13G) [inline onair/src/run_scripts/sim.py:42]: swap

**Review by Evana13G — COMMENTED** [review]
Small tweaks

**Comment by Evana13G** [comment]
@cfirth-nasa Nice job :D I ran a pedantic review, happy to discuss if there are questions. I'll merge after these small tweaks. Note, some of the code changes will have ramifications in the test files. Thank youuuu!

**Comment by cfirth-nasa** [comment]
Will have time to work on this tomorrow - functionality is there but unit tests need work. Leaving up in case someone (Alan) can get to it before I can

**Pull request merged**